### PR TITLE
Add a 'download file' link in footer

### LINKED ISSUE #287: Ability to get raw file
Sometimes I'd like to take a diff of the same file on different versions. If there were a way to get the raw file, it would be easier to do this with elixir as source of the file. A parameter like `?raw` (or `?raw=1`) would already be enough in many cases - e.g. https://elixir.bootlin.com/linux/v6.7.10/source/fs/init.c?raw

### PR DESCRIPTION
**Pull request by fstachura**
Closes #287

### CODE REVIEW

File: elixir/web.py [hunk 1]
```diff
@@ -92,8 +92,7 @@ def on_get(self, req, resp, project, version, path):
             version = parse.quote(query.query('latest'))
             raise falcon.HTTPFound(stringify_source_path(project, version, path))
 
-        resp.content_type = falcon.MEDIA_HTML
-        resp.status, resp.text = generate_source_page(req.context, query, project, version, path)
+        return generate_source_page(req, resp, query, project, version, path)
```
Comment (tleb) [inline elixir/web.py:95]: This change does not look relevant to the raw file feature.

File: elixir/web.py [hunk 2]
```diff
@@ -380,16 +348,60 @@ def generate_source_page(ctx, q, project, version, path):
...
-    return (status, template.render(data))
+# Generates response (status code and optionally HTML) of the `source` route
+# ctx: RequestContext
+# q: Query object
+# parsed_path: ParsedSourcePath
+def generate_source_page(req, resp, q, project, version, path):
+    version_unquoted = parse.unquote(version)
+    source_base_url = f'/{ project }/{ version }/source'
+
+    type = q.query('type', version_unquoted, path)
```
Comment (tleb) [inline elixir/web.py:366]: Can we avoid the `type` name for a variable? It has a specific meaning in Python, so we overload a default function.

File: elixir/web.py [hunk 3]
```diff
@@ -380,16 +348,60 @@ def generate_source_page(ctx, q, project, version, path):
...
+        })
+    elif type == 'blob':
+        if 'raw' in req.params and req.params['raw'] != '0':
+            resp.text = q.get_file_raw(parse.unquote(version), path)
+
+            resp.downloadable_as = path.split('/')[-1]
+            resp.content_type = 'application/octet-stream'
+            # 24 hours
+            resp.cache_control = ('max-age=86400',)
+            resp.headers['Content-Security-Policy'] = "sandbox; default-src 'none'"
```
Comment (tleb) [inline elixir/web.py:389]: There seems to be some behavior changes that have no relation to the raw file feature that are hidden in this seemingly simple change.

File: static/style.css [hunk 4]
```diff
@@ -404,6 +404,9 @@ h2 {
           box-shadow: 0 0 0 1px #666;
 }
 
+#file-download-link {
+  margin-left: 1rem;
```
Comment (tleb) [inline static/style.css:408]: Is indentation in line with the rest of the file?

File: templates/source.html [hunk 5]
```diff
@@ -9,3 +9,9 @@
     {{ code }}
 </div>
 {% endblock %}
+
+{% block footer %}
+<span id="file-download-link">
+    <a href="?raw">Download file</a>
```
Comment (tleb) [inline templates/source.html:15]: Is `?raw` standard HTTP? It means nicer URL but we should avoid it if `?raw=1` is required by the spec.

File: elixir/web.py (hunk 3 above)
Comment (fstachura) [reply elixir/web.py:389]: This is only executed if the raw parameter is present. The purpose of all of this is to make sure that the file is downloaded, not opened (and definitely not as something that can be interpreted by the browser) and that it's cached for a reasonable amount of time - raw files shouldn't change much.

Regarding Content-Security-Policy, many other websites do something similar, for example see headers on Kernel cgit: https://git.kernel.org/pub/scm/docs/kernel/ksmap.git/plain/sample.yaml
(although others usually use it also outside of raw files. maybe CSP is not going to be that useful here, but I think it won't hurt).
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy

File: elixir/web.py (hunk 1 above)
Comment (fstachura) [reply elixir/web.py:95]: Without this, generate_source_page cannot add headers. Unless I do some workaround

File: templates/source.html (hunk 5 above)
Comment (fstachura) [reply templates/source.html:15]: I couldn't find anything good to quote from a standard. From what I have read, HTTP does not actually care how query parameters are encoded. JS and HTML on the other hand...

Apparently sometimes query params are expected to be  www-form-urlencoded, which allows keys without a value: https://url.spec.whatwg.org/#urlencoded-parsing
And JS URLSearchParams expect it to be www-form-urlencoded: https://url.spec.whatwg.org/#interface-urlsearchparams

```
> new URL('https://example.com/?z=x&a&b&c').searchParams
URLSearchParams(4) { z → "x", a → "", b → "", c → "" }
```
Although 
```
> let url = new URL('https://example.com/?z=x&a')
> url.searchParams.set('test', '')
> url.search
"?z=x&a=&test=" 
```

Random stackoverflow question: https://stackoverflow.com/questions/4557387/is-a-url-query-parameter-valid-if-it-has-no-value
On framework side this gets confusing: https://stackoverflow.com/questions/3130541/access-query-string-parameters-with-no-values-in-asp-net
In go you apparently can't even build such URLs: https://old.reddit.com/r/golang/comments/10hud2z/is_there_a_way_to_build_a_query_string_with_no/
But Falcon seems to support it.

So I don't know. I think it's standard but just niche. It should work fine, but I'm open to using `=1` or maybe even changing the URL format completely.

File: static/style.css (hunk 4 above)
Comment (fstachura) [reply static/style.css:408]: Yep

```
.go-top:hover::before {
  color: #fff;
  background: #666;
  -webkit-box-shadow: 0 0 0 1px #666;
          box-shadow: 0 0 0 1px #666;
}

#file-download-link {
  margin-left: 1rem;
}
```

File: templates/source.html (hunk 5 above)
Comment (tleb) [reply templates/source.html:15]: It is fine as is then. It makes for nicer URLs.

File: elixir/web.py (hunk 3 above)
Comment (tleb) [reply elixir/web.py:389]: Ah I missed that. Please expand (in commit message or code comment) on why this specific CSP value was picked (sandbox + default-src none). I am not understanding the sandbox directive without any value, even when briefly reading [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/sandbox) or [the spec](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#attr-iframe-sandbox).

File: elixir/web.py (hunk 1 above)
Comment (tleb) [reply elixir/web.py:95]: The rework of `generate_source_page` could be done in a separate commit. Else the "Add support for 'raw' parameter for source files" commit message is a lie and should be corrected.

File: elixir/web.py (hunk 3 above)
Comment (fstachura) [reply elixir/web.py]: I added more info in a commit message

File: elixir/web.py (hunk 1 above)
Comment (fstachura) [reply elixir/web.py]: I totally reworked the patch in and I think this is not relevant now

**Comment by fstachura** [comment]
I also have a version that returns the contents of the file without triggering a download. Maybe this is actually more useful, or maybe Elixir should have links for both. 

https://github.com/fstachura/elixir/tree/raw-file

**Comment by tleb** [comment]
> I also have a version that returns the contents of the file without triggering a download. Maybe this is actually more useful, or maybe Elixir should have links for both.

Thanks for the proposal, but I'd say the initial behavior is better and more intuitive to most users.

Merged as 8c5e12fffd9782b91f99036e4972d6d9f1b4b27f, thanks!
